### PR TITLE
Enable correct resumption from the end of an epoch

### PR DESCRIPTION
**Pull request by snarayan21**
## Description of changes:

This PR enables correct resumption from the end of an epoch. For example, suppose the dataset has 512 samples and I want to resume from sample 512. The intended behavior should be to detect that we are at the end of the epoch and start a new one. There were two small bugs preventing this from happening correctly:

1. The partition algorithms (`orig` and `relaxed`) both threw an error when `num_samples <= drop_first` instead of `num_samples < drop_first`. This means that if we wanted to create a sample partition of 512 samples but drop the first 512 samples, we would throw an error. This is not the right behavior -- we should return an empty sample partition instead.
2. Writing an empty sample partition array to shared memory was not possible since we did not consider that edge case. This has been addressed.

Manual run that had earlier failed, succeeding with these changes: `resumption-issue-testing-RnnDR4`

Unit tests have been added for these cases.

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [ ] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [ ] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [ ] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/dataset.py [hunk 1]
```diff
@@ -922,14 +922,16 @@ def resample_streams(
         sample_ids = np.concatenate(sample_ids).astype(np.int64)
         return shuffle_units, sample_ids
 
-    def _share_work(self, sample_ids: NDArray[np.int64]) -> Tuple[SharedMemory, SharedMemory]:
+    def _share_work(self,
+                    sample_ids: NDArray[np.int64]) -> Tuple[SharedMemory, Optional[SharedMemory]]:
```
Comment (mvpatel2000) [inline streaming/base/dataset.py:926]:
```suggestion
                    sample_ids: NDArray[np.int64],) -> Tuple[SharedMemory, Optional[SharedMemory]]:
```
cleaner formatting with trailing ,

File: streaming/base/partition/relaxed.py [hunk 2]
```diff
@@ -65,6 +65,7 @@ def get_partitions_relaxed(num_samples: int,
         return get_partitions_orig(num_samples, num_canonical_nodes, num_physical_nodes,
                                    ranks_per_node, workers_per_rank, batch_size, drop_first)
     else:
+        print('WE HERE')
```
Comment (mvpatel2000) [inline streaming/base/partition/relaxed.py:68]:
```suggestion
```

File: streaming/base/partition/relaxed.py [hunk 3]
```diff
@@ -82,6 +83,7 @@ def get_partitions_relaxed(num_samples: int,
         partition = get_partitions_orig(num_samples, num_canonical_nodes, initial_physical_nodes,
                                         ranks_per_node, workers_per_rank, initial_batch_size,
                                         drop_first)
+        print('ORIG PARTITION SHAPE: ', partition.shape)
```
Comment (mvpatel2000) [inline streaming/base/partition/relaxed.py:86]:
```suggestion
```

**Review by mvpatel2000 — APPROVED** [review]

File: streaming/base/partition/relaxed.py (hunk 2 above)
Comment (snarayan21) [reply streaming/base/partition/relaxed.py:68]: whoops lol

File: streaming/base/dataset.py (hunk 1 above)
Comment (snarayan21) [reply streaming/base/dataset.py]: done
